---
title: "tetrapodtraits"
author: "Jimmy Choi"
date: "2024-10-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "C:/Users/jimmy/OneDrive/Documents")
library(ape)
library(plyr)
library(ggplot2)
library(tibble)
library(dplyr)
library(geiger)
library(effsize)
library(phylolm)
library(sensiPhy)
library(chromePlus)
library(diversitree)
```

```{r}
tetrapodtraits <- read.csv(file = "TetrapodTraits_1.0.0.csv", stringsAsFactors = FALSE)
tetrapodtraits.herp <- tetrapodtraits[which(tetrapodtraits$Family == "Herpestidae"),]
tetrapodtraits.herp
```

```{r}
mammal.phylo <- read.nexus(file = "bettermammalphylogeny.tre")
short.tip.labels <- gsub('^([^_]+_[^_]+).*', '\\1',mammal.phylo$tip.label)
short.tip.labels <- gsub('_',' ',short.tip.labels)
mammal.phylo$tip.label <- short.tip.labels
## fix taxonomy to be congruous with MDD at some point but it works for now... bleh
```

```{r}
herp.node <- getMRCA(mammal.phylo,tetrapodtraits.herp$Scientific.Name) ## do i know if this tree is node dated or tip dated?
herpestidae.tree <- extract.clade(mammal.phylo,herp.node)
herpestidae.tree$tip.label
herpestidae.tree$tip.label <- mapvalues(herpestidae.tree$tip.label, from=c("Herpestes brachyurus","Herpestes edwardsii","Herpestes fuscus","Herpestes semitorquatus","Herpestes smithii","Herpestes urva","Herpestes vitticollis","Herpestes javanicus","Herpestes naso"), to=c("Urva brachyura","Urva edwardsii","Urva fusca","Urva semitorquata","Urva smithii","Urva urva","Urva vitticollis","Urva javanica","Xenogale naso"))
tetrapodtraits.herp$Scientific.Name <- mapvalues(tetrapodtraits.herp$Scientific.Name, from=c("Herpestes brachyurus","Herpestes edwardsii","Herpestes fuscus","Herpestes semitorquatus","Herpestes smithii","Herpestes urva","Herpestes vitticollis","Herpestes javanicus","Herpestes naso"), to=c("Urva brachyura","Urva edwardsii","Urva fusca","Urva semitorquata","Urva smithii","Urva urva","Urva vitticollis","Urva javanica","Xenogale naso"))
```

```{r}
sex.reference <- read.csv(file = "sexreferencedf.csv")
tetrapodtraits.herp <- add_column(tetrapodtraits.herp,sex="",.before = 1)
tetrapodtraits.herp$sex <- sex.reference[match(tetrapodtraits.herp$Scientific.Name, sex.reference$species),2]
tetrapodtraits.herp <- add_column(tetrapodtraits.herp,AbsoluteLatitude="",.after = 63)
tetrapodtraits.herp$AbsoluteLatitude <- abs(tetrapodtraits.herp$Latitude)
tetrapodtraits.herp <- tetrapodtraits.herp[!sapply(tetrapodtraits.herp, function(x) all(x == ""))]
tetrapodtraits.herp
```


```{r}
tetrapodtraits.herp <- add_column(tetrapodtraits.herp,XYorvariant="",.before = 1)
tetrapodtraits.herp$XYorvariant <- tetrapodtraits.herp$sex
tetrapodtraits.herp$XYorvariant <- mapvalues(tetrapodtraits.herp$XYorvariant, from = c("XY","YAfusion"), to = c(0,1))
tetrapodtraits.herp$XYorvariant <- as.numeric(tetrapodtraits.herp$XYorvariant)
## XY = 0, variant = 1
tetrapodtraits.herp <- tetrapodtraits.herp[!sapply(tetrapodtraits.herp, function(x) all(x == ""))]
tetrapodtraits.herp
```

```{r}
rownames(tetrapodtraits.herp) <- unlist(tetrapodtraits.herp$Scientific.Name)
var_names <- c("RangeSize","AbsoluteLatitude","AnnuMeanTemp","AnnuPrecip","TempSeasonality","PrecipSeasonality","Elevation") 
tetrapodtraits.herp[,var_names] <- sapply(tetrapodtraits.herp[,var_names], as.numeric)
tetrapodtraits.herp$XYorvariant <- as.numeric(tetrapodtraits.herp$XYorvariant)
cohens_d <- c()
log_reg <- c()
## log_reg2 <- c()
## alpha_values <- c()

for (var in var_names) {
    
x <- cohen.d(na.omit(tetrapodtraits.herp[tetrapodtraits.herp$sex=='YAfusion',][[var]]), na.omit(tetrapodtraits.herp[tetrapodtraits.herp$sex=='XY',][[var]]))
cohens_d <- c(cohens_d, x$estimate)

y <- tetrapodtraits.herp[[var]]
names(y) <- unlist(tetrapodtraits.herp$Scientific.Name)
group <- tetrapodtraits.herp$XYorvariant
#names(group) <- unlist(tetrapodtraits.herp$Scientific.Name)
test <- phyloglm(group ~ y, phy = keep.tip(herpestidae.tree, tetrapodtraits.herp[!(is.na(tetrapodtraits.herp[[var]])),]$Scientific.Name), data = tetrapodtraits.herp, method = c("logistic_MPLE"), boot = 10)
## test2<-binaryPGLMM(group ~ y, phy = keep.tip(herpestidae.tree, tetrapodtraits.herp[!(is.na(tetrapodtraits.herp[[var]])),]$Scientific.Name), data = tetrapodtraits.herp,tol.pql = 1e-6,maxit.pql = 400,maxit.reml = 200)
log_reg <- rbind(log_reg, summary(test))
## log_reg2 <- rbind(log_reg2, test2)
## alpha_values <- rbind(alpha_values, test)
print(test$alphaWarn)
}
## vectorization?, pivot table? tidyr, dplyr
## log_reg_p_values <- log_reg[1:7,2]
## log_reg_p_values[[1]][12] so on so forth, use append() to add to list of p values
## herp_log_reg_p
## also check out PhyR package? PGLMM function (phylo generalized linear mixed model)
## divide all branch lengths by the mean of all branch lengths for phyloglm() analysis
## see if optimization problem still happens w/ reduced trees (trees w only known sex)
## drop Sorex kozlovi/label as a fusion?
```

```{r}
color <- recode(tetrapodtraits.herp$sex, "XY" = "black", "YAfusion" = "red")
herp.mycol <- with(tetrapodtraits.herp, as.character(color[match(herpestidae.tree$tip.label, tetrapodtraits.herp$Scientific.Name)]))
plot(herpestidae.tree, no.margin = TRUE, cex = 0.5, tip.color = herp.mycol)
```

```{r}
tetrapodtraits.phyllostom <- tetrapodtraits[which(tetrapodtraits$Family == "Phyllostomidae"),]
tetrapodtraits.phyllostom
## tetrapodtraits.phyllostom$Scientific.Name <- mapvalues(tetrapodtraits.phyllostom$Scientific.Name, from=c("Artibeus toltecus","Artibeus phaeotis","Artibeus rosenbergi","Artibeus watsoni","Artibeus aztecus","Artibeus glaucus","Artibeus cinereus","Artibeus gnomus","Artibeus bogotensis","Artibeus ravus","Artibeus anderseni","Gardnerycteris crenulatum","Vampyriscus nymphaea","Carollia brevicaudum"), to=c("Dermanura tolteca","Dermanura phaeotis","Dermanura rosenbergi","Dermanura watsoni","Dermanura azteca","Dermanura glauca","Dermanura cinerea","Dermanura gnoma","Dermanura bogotensis","Dermanura rava","Dermanura anderseni","Gardnerycteris crenulata","Vampyriscus nymphaeus","Carollia brevicauda"))
## mammal.phylo$tip.label <- mapvalues(mammal.phylo$tip.label, from=c("Diaemus youngi","Mimon koepckeae","Dermanura toltecus","Dermanura rosenbergii","Dermanura aztecus","Dermanura glaucus","Dermanura cinereus","Dermanura gnomus","Vampyressa bidens","Vampyressa brocki","Vampyressa nymphaea","Mimon crenulatum","Lophostoma aequatorialis","Lophostoma silvicolum"), to=c("Diaemus youngii","Gardnerycteris koepckeae","Dermanura tolteca","Dermanura rosenbergi","Dermanura azteca","Dermanura glauca","Dermanura cinerea","Dermanura gnoma","Vampyriscus bidens","Vampyriscus brocki","Vampyriscus nymphaeus","Gardnerycteris crenulata","Lophostoma occidentale","Lophostoma silvicola"))
## mammal.phylo.long <- read.nexus(file = "bettermammalphylogeny.tre")
## phyllo.taxa <- grep("PHYLLOSTOMIDAE", mammal.phylo.long$tip.label, value = TRUE)
## phyllo.node <- getMRCA(mammal.phylo.long, phyllo.taxa)
## phyllo.node = 7781
phyllostom.tree <- extract.clade(mammal.phylo,7781)
```


```{r}
setdiff(tetrapodtraits.phyllostom$Scientific.Name, phyllostom.tree$tip.label)
## D_draculae -> D. draculae
```


```{r}
## problem.taxa <- tetrapodtraits.phyllostom[is.na(tetrapodtraits.phyllostom$sex),]
## problem.taxa$Scientific.Name

tetrapodtraits.phyllostom$Scientific.Name <- mapvalues(tetrapodtraits.phyllostom$Scientific.Name, from=c("Dermanura aztecus", "Dermanura cinereus", "Dermanura glaucus", "Dermanura gnomus", "Dermanura rosenbergii", "Dermanura toltecus", "Lophostoma aequatorialis", "Lophostoma silvicolum", "Mimon crenulatum", "Mimon koepckeae", "Vampyressa bidens", "Vampyressa brocki", "Vampyressa nymphaea", "Desmodus_draculae", "Diaemus youngi"), to=c("Dermanura azteca", "Dermanura cinerea", "Dermanura glauca", "Dermanura gnoma", "Dermanura rosenbergi", "Dermanura tolteca", "Lophostoma occidentale", "Lophostoma silvicola", "Gardnerycteris crenulata", "Gardnerycteris koepckeae", "Vampyriscus bidens", "Vampyriscus brocki", "Vampyriscus nymphaeus", "Desmodus draculae", "Diaemus youngii"))
phyllostom.tree$tip.label <- mapvalues(phyllostom.tree$tip.label, from=c("Dermanura aztecus", "Dermanura cinereus", "Dermanura glaucus", "Dermanura gnomus", "Dermanura rosenbergii", "Dermanura toltecus", "Lophostoma aequatorialis", "Lophostoma silvicolum", "Mimon crenulatum", "Mimon koepckeae", "Vampyressa bidens", "Vampyressa brocki", "Vampyressa nymphaea", "Diaemus youngi"), to=c("Dermanura azteca", "Dermanura cinerea", "Dermanura glauca", "Dermanura gnoma", "Dermanura rosenbergi", "Dermanura tolteca", "Lophostoma occidentale", "Lophostoma silvicola", "Gardnerycteris crenulata", "Gardnerycteris koepckeae", "Vampyriscus bidens", "Vampyriscus brocki", "Vampyriscus nymphaeus", "Diaemus youngii"))
```


```{r}
tetrapodtraits.phyllostom <- add_column(tetrapodtraits.phyllostom,sex="",.before = 1)
tetrapodtraits.phyllostom$sex <- sex.reference[match(tetrapodtraits.phyllostom$Scientific.Name, sex.reference$species),2]
tetrapodtraits.phyllostom <- add_column(tetrapodtraits.phyllostom,AbsoluteLatitude="",.after = 63)
tetrapodtraits.phyllostom$AbsoluteLatitude <- abs(tetrapodtraits.phyllostom$Latitude)
tetrapodtraits.phyllostom <- add_column(tetrapodtraits.phyllostom,XYorvariant="",.before = 1)
tetrapodtraits.phyllostom$XYorvariant <- mapvalues(tetrapodtraits.phyllostom$sex, from = c("XY","XAfusion","XAYAfusion"), to = c(0,1,1))
## XY = 0, variant = 1
tetrapodtraits.phyllostom$XYorvariant <- as.numeric(tetrapodtraits.phyllostom$XYorvariant)
tetrapodtraits.phyllostom <- tetrapodtraits.phyllostom[!sapply(tetrapodtraits.phyllostom, function(x) all(x == ""))]
tetrapodtraits.phyllostom
```


```{r}
ggplot(tetrapodtraits.phyllostom, aes(x = reorder(Scientific.Name, RangeSize), y = RangeSize, fill = sex)) + 
geom_bar(stat="identity") +
scale_x_discrete(guide = guide_axis(angle = 45))
```

```{r}
ggplot(tetrapodtraits.phyllostom, aes(x = reorder(Scientific.Name, AbsoluteLatitude), y = AbsoluteLatitude, fill = sex)) + 
geom_bar(stat="identity") +
scale_x_discrete(guide = guide_axis(angle = 45)) +
scale_fill_manual(values=c("#FF0000", "#E69F00", "#56B4E9"))
```

```{r}
rownames(tetrapodtraits.phyllostom) <- unlist(tetrapodtraits.phyllostom$Scientific.Name)
var_names <- c("RangeSize","AbsoluteLatitude","AnnuMeanTemp","AnnuPrecip","TempSeasonality","PrecipSeasonality","Elevation")
tetrapodtraits.phyllostom[,var_names] <- sapply(tetrapodtraits.phyllostom[,var_names], as.numeric)
cohens_d <- c()
log_reg2 <- c()

for (var in var_names) {
    
x <- cohen.d(na.omit(tetrapodtraits.phyllostom[tetrapodtraits.phyllostom$XYorvariant==0,][[var]]), na.omit(tetrapodtraits.phyllostom[tetrapodtraits.phyllostom$XYorvariant==1,][[var]]))
cohens_d <- c(cohens_d, x$estimate)

y <- tetrapodtraits.phyllostom[[var]]
names(y) <- unlist(tetrapodtraits.phyllostom$Scientific.Name)
group <- factor(tetrapodtraits.phyllostom$XYorvariant)
names(group) <- unlist(tetrapodtraits.phyllostom$Scientific.Name) 
test <- phyloglm(group ~ y, phy = keep.tip(phyllostom.tree, tetrapodtraits.phyllostom[!(is.na(tetrapodtraits.phyllostom[[var]])),]$Scientific.Name), data = tetrapodtraits.phyllostom, method = c("logistic_MPLE"), boot = 10)
log_reg2 <- rbind(log_reg2, summary(test))
print(test$alphaWarn)
}
## check each variable in each clade for corresponding warning messages
## phyllo_log_reg_p
```

```{r}
color <- recode(tetrapodtraits.phyllostom$sex, "XY" = "black", "XAYAfusion" = "purple", "XAfusion" = "blue")
phyllo.mycol <- with(tetrapodtraits.phyllostom, as.character(color[match(phyllostom.tree$tip.label, tetrapodtraits.phyllostom$Scientific.Name)]))
plot(phyllostom.tree, no.margin = TRUE, cex = 0.3, tip.color = phyllo.mycol)
```


```{r}
tetrapodtraits.sorex <- tetrapodtraits[which(tetrapodtraits$Family == "Soricidae"),]
tetrapodtraits.sorex
```

```{r}
sorex.node <- getMRCA(mammal.phylo, tetrapodtraits.sorex$Scientific.Name)
sorex.tree <- extract.clade(mammal.phylo, sorex.node)
sorex.tree
```

```{r}
setdiff(sorex.tree$tip.label, tetrapodtraits.sorex$Scientific.Name)
```

```{r}
variant.mamm <- read.csv(file = "variant_mammals.csv", header=T)
variant.sorex <- variant.mamm$X.2[77:86]
variant.sorex
```

```{r}
tetrapodtraits.sorex <- add_column(tetrapodtraits.sorex,sex="",.before = 1)
tetrapodtraits.sorex$sex <- sex.reference[match(tetrapodtraits.sorex$Scientific.Name, sex.reference$species),2]
tetrapodtraits.sorex <- add_column(tetrapodtraits.sorex,AbsoluteLatitude="",.after = 63)
tetrapodtraits.sorex$AbsoluteLatitude <- abs(tetrapodtraits.sorex$Latitude)
tetrapodtraits.sorex <- add_column(tetrapodtraits.sorex,XYorvariant="",.before = 1)
tetrapodtraits.sorex$XYorvariant <- mapvalues(tetrapodtraits.sorex$sex, from = c("XY","XAfusion"), to = c(0,1))
## XY = 0, variant = 1
tetrapodtraits.sorex$XYorvariant <- as.numeric(tetrapodtraits.sorex$XYorvariant)
tetrapodtraits.sorex <- tetrapodtraits.sorex[!sapply(tetrapodtraits.sorex, function(x) all(x == ""))]
tetrapodtraits.sorex
```


```{r}
rownames(tetrapodtraits.sorex) <- unlist(tetrapodtraits.sorex$Scientific.Name)
var_names <- c("RangeSize","AbsoluteLatitude","AnnuMeanTemp","AnnuPrecip","TempSeasonality","PrecipSeasonality","Elevation")
cohens_d <- c()
log_reg3 <- c()

for (var in var_names) {
    
x <- cohen.d(na.omit(tetrapodtraits.sorex[tetrapodtraits.sorex$XYorvariant==1,][[var]]), na.omit(tetrapodtraits.sorex[tetrapodtraits.sorex$XYorvariant==0,][[var]]))
cohens_d <- c(cohens_d, x$estimate)

y <- tetrapodtraits.sorex[[var]]
names(y) <- unlist(tetrapodtraits.sorex$Scientific.Name)
group <- factor(tetrapodtraits.sorex$XYorvariant)
names(group) <- unlist(tetrapodtraits.sorex$Scientific.Name) 
test <- phyloglm(group ~ y, phy = keep.tip(sorex.tree, tetrapodtraits.sorex[!(is.na(tetrapodtraits.sorex[[var]])),]$Scientific.Name), data = tetrapodtraits.sorex, method = c("logistic_MPLE"), boot = 10)
log_reg3 <- rbind(log_reg3, summary(test))
print(test$alphaWarn)
    
}
## maybe actually use cohens_d at some point for analysis?
## sorex_log_reg_p
```


```{r}
ggplot(tetrapodtraits.sorex, aes(x = reorder(tetrapodtraits.sorex$Scientific.Name, AbsoluteLatitude), y = AbsoluteLatitude, fill = sex)) + 
geom_bar(stat="identity") +
scale_x_discrete(guide = guide_axis(angle = 45)) +
theme(axis.text=element_text(size=2.5)) +
labs(x = "Soricidae species sorted by AbsoluteLat")
## map species ranges, get other ecovariables from the same dataset as tetrapodtraits integrated, read more lit!
```

```{r}
ggplot(tetrapodtraits.sorex, aes(x = reorder(tetrapodtraits.sorex$Scientific.Name, RangeSize), y = RangeSize, fill = sex)) + 
geom_bar(stat="identity") +
scale_x_discrete(guide = guide_axis(angle = 45))
```

```{r}
herp.multiphylo <- read.nexus(file = "herpoutput.nex")
herp.multiphylo
```

```{r}
phy.replace.herp <- function(tree) {
  tree$tip.label <- sub("_", " ", tree$tip.label)
  tree$tip.label <- mapvalues(tree$tip.label, from=c("Herpestes brachyurus","Herpestes edwardsii","Herpestes fuscus","Herpestes semitorquatus","Herpestes smithii","Herpestes urva","Herpestes vitticollis","Herpestes javanicus","Herpestes naso"), to=c("Urva brachyura","Urva edwardsii","Urva fusca","Urva semitorquata","Urva smithii","Urva urva","Urva vitticollis","Urva javanica","Xenogale naso"))
  return(tree)
}
```

```{r}
herp.multiphylo <- lapply(herp.multiphylo, phy.replace.herp)
class(herp.multiphylo) <- "multiPhylo"
```

```{r}
# reprex() - a tidyverse function
# sink() - base R
var_names <- c("RangeSize","AbsoluteLatitude","AnnuMeanTemp","AnnuPrecip","TempSeasonality","PrecipSeasonality","Elevation")
sensiphy_analyses <- c()

for (var in var_names) {

herp.sensiphy <- tree_phyglm(XYorvariant ~ tetrapodtraits.herp[[var]], data = tetrapodtraits.herp, phy = herp.multiphylo, n.tree = 1000)
sensiphy_analyses <- rbind(sensiphy_analyses, summary(herp.sensiphy))

}
```

```{r}
herp.sensiphy <- tree_phyglm(XYorvariant ~ Elevation, data = tetrapodtraits.herp, phy = herp.multiphylo, n.tree = 1000)
herp.sensiphy
## assumption that all non-variant species are XY?
```

```{r}
phyllostom.multiphylo <- read.nexus(file = "phyllostomoutput.nex")
phyllostom.multiphylo
```

```{r}
phy.replace.phyllostom <- function(tree) {
  tree$tip.label <- sub("_", " ", tree$tip.label)
  tree$tip.label <- mapvalues(tree$tip.label, from=c("Dermanura aztecus", "Dermanura cinereus", "Dermanura glaucus", "Dermanura gnomus", "Dermanura rosenbergii", "Dermanura toltecus", "Lophostoma aequatorialis", "Lophostoma silvicolum", "Mimon crenulatum", "Mimon koepckeae", "Vampyressa bidens", "Vampyressa brocki", "Vampyressa nymphaea", "Diaemus youngi"), to=c("Dermanura azteca", "Dermanura cinerea", "Dermanura glauca", "Dermanura gnoma", "Dermanura rosenbergi", "Dermanura tolteca", "Lophostoma occidentale", "Lophostoma silvicola", "Gardnerycteris crenulata", "Gardnerycteris koepckeae", "Vampyriscus bidens", "Vampyriscus brocki", "Vampyriscus nymphaeus", "Diaemus youngii"))
  return(tree)
}
```

```{r}
phyllostom.multiphylo <- lapply(phyllostom.multiphylo, phy.replace.phyllostom)
class(phyllostom.multiphylo) <- "multiPhylo"
```

```{r}
phyllostom.sensiphy.elevation <- tree_phyglm(XYorvariant ~ Elevation, data = tetrapodtraits.phyllostom, phy = phyllostom.multiphylo, n.tree = 1000)
summary(phyllostom.sensiphy.elevation)
```

```{r}
sorex.multiphylo <- read.nexus(file = "sorexoutput.nex")
sorex.multiphylo
```

```{r}
phy.replace.sorex <- function(tree) {
  tree$tip.label <- sub("_", " ", tree$tip.label)
  return(tree)
}
```

```{r}
sorex.multiphylo <- lapply(sorex.multiphylo, phy.replace.sorex)
class(sorex.multiphylo) <- "multiPhylo"
```

```{r}
sorex.sensiphy <- tree_phyglm(XYorvariant ~ Elevation, data = tetrapodtraits.sorex, phy = sorex.multiphylo, n.tree = 1000)
summary(sorex.sensiphy)
```

```{r}
color <- recode(tetrapodtraits.sorex$sex, "XY" = "black", "XAfusion" = "blue")
sorex.mycol <- with(tetrapodtraits.sorex, as.character(color[match(sorex.tree$tip.label, tetrapodtraits.sorex$Scientific.Name)]))
plot(sorex.tree, no.margin = TRUE, cex = 0.2, tip.color = sorex.mycol)
```

```{r}
sorex.variant.node <- getMRCA(sorex.tree, c("Sorex antinorii","Sorex araneus","Sorex arcticus","Sorex asper","Sorex coronatus","Sorex daphaenodon","Sorex granarius","Sorex maritimensis","Sorex satunini","Sorex tundrensis"))
sorex.variant.tree <- extract.clade(sorex.tree, sorex.variant.node)
```

```{r}
##plotTree(sorex.variant.tree, ftype="i", fsize=1, lwd=1)
##add.arrow(sorex.variant.tree, tip = c("Sorex antinorii","Sorex araneus","Sorex arcticus","Sorex asper","Sorex coronatus","Sorex daphaenodon","Sorex granarius","Sorex maritimensis","Sorex satunini","Sorex tundrensis"),lwd=0.5,arrl=0.1,col="blue",offset=2)
sorex.variant.col <- with(tetrapodtraits.sorex, as.character(color[match(sorex.variant.tree$tip.label, tetrapodtraits.sorex$Scientific.Name)]))
plot(sorex.variant.tree, no.margin = TRUE, cex = 1, tip.color = sorex.variant.col)
## setting Sorex kozlovi as XY is an unsafe assumption, should do something about this
```

```{r}
karyotype.data <- read.csv(file = "chroms.csv")
karyotype.data
## check out jon's review paper too, only analyze species with known karyotype data and compare to all species (w assumptions)? trait evolution modeling?
```

```{r}
karyotype.data$Species <- gsub('_', ' ', karyotype.data$Species)
karyotype.data
```

```{r}
tetrapodtraits.herp <- add_column(tetrapodtraits.herp, perc.acro = "")
tetrapodtraits.herp$perc.acro <- karyotype.data[match(tetrapodtraits.herp$Scientific.Name, karyotype.data$Species),5]
tetrapodtraits.herp <- tetrapodtraits.herp[!sapply(tetrapodtraits.herp, function(x) all(x == ""))]
tetrapodtraits.herp
```

```{r}
ggplot(tetrapodtraits.herp, aes(x = reorder(tetrapodtraits.herp$Scientific.Name, perc.acro), y = perc.acro, fill = sex)) +
geom_bar(stat="identity") +
scale_x_discrete(guide = guide_axis(angle = 45))
```

```{r}
## tetrapodtraits.herp.variant
herp.no.acro.data <- tetrapodtraits.herp$Scientific.Name[which(is.na(tetrapodtraits.herp$perc.acro))]
```

```{r}
jon.karyotype.data <- read.csv(file = "family_level_known_karyotypes.csv")
jon.karyotype.data
```

```{r}
jon.karyotype.data$mdd_species <- gsub('_', ' ', jon.karyotype.data$mdd_species)
jon.karyotype.data
```

```{r}
jon.karyotype.data.herps <- jon.karyotype.data[jon.karyotype.data$family == "HERPESTIDAE",]
jon.karyotype.data.herps$mdd_species[which(is.na(jon.karyotype.data.herps$sex_chr))]
```


```{r}
tetrapodtraits.herp <- add_column(tetrapodtraits.herp, sex.imputed = "", .before = 2)
tetrapodtraits.herp$sex.imputed <- tetrapodtraits.herp$XYorvariant
## sex.imputed == 1 means we made XY assumption, sex.imputed == 0 means we have source
tetrapodtraits.herp$sex.imputed <- mapvalues(tetrapodtraits.herp$XYorvariant, from = c(1,0), to = c(0,"XY"))
## XY because we're not sure yet, but should be safe to say if it's XX/XY and has perc.acro value and/or a citation in jon.karyotype.data, there's a source for its XX/XYness
tetrapodtraits.herp <- tetrapodtraits.herp[!sapply(tetrapodtraits.herp, function(x) all(x == ""))]
tetrapodtraits.herp
```

```{r}
XY.with.blackmon.data <- intersect(tetrapodtraits.herp$Scientific.Name[tetrapodtraits.herp$sex.imputed == "XY"], tetrapodtraits.herp$Scientific.Name[which(!is.na(tetrapodtraits.herp$perc.acro))])
XY.with.blackmon.data
```

```{r}
tetrapodtraits.herp$sex.imputed[which(tetrapodtraits.herp$Scientific.Name %in% XY.with.blackmon.data)] <- 0
tetrapodtraits.herp
```

```{r}
XY.with.jon.data <- intersect(jon.karyotype.data.herps$mdd_species[which(!is.na(jon.karyotype.data.herps$sex_chr))], tetrapodtraits.herp$Scientific.Name[tetrapodtraits.herp$sex.imputed == "XY"])
XY.with.jon.data
```

```{r}
tetrapodtraits.herp$sex.imputed[which(tetrapodtraits.herp$Scientific.Name %in% XY.with.jon.data)] <- 0
tetrapodtraits.herp
```

```{r}
tetrapodtraits.herp$sex.imputed <- mapvalues(tetrapodtraits.herp$sex.imputed, from=c("XY"), to=c(1))
tetrapodtraits.herp
```

```{r}
tetrapodtraits.herp$perc.acro[4] <- tetrapodtraits.herp$Scientific.Name[4]
tetrapodtraits.herp$perc.acro[28] <- tetrapodtraits.herp$Scientific.Name[28]
tetrapodtraits.herp$perc.acro[34] <- tetrapodtraits.herp$Scientific.Name[34]
tetrapodtraits.herp$perc.acro <- mapvalues(tetrapodtraits.herp$perc.acro, from = c("Bdeogale nigripes", "Ichneumia albicauda", "Suricata suricatta"), to = c(0.16666667, 0.16666667, 0.16666667))
tetrapodtraits.herp$perc.acro
```

```{r}
tetrapodtraits.herp.known.karyo <- tetrapodtraits.herp[!is.na(tetrapodtraits.herp$perc.acro),]
tetrapodtraits.herp.known.karyo$perc.acro <- as.numeric(tetrapodtraits.herp.known.karyo$perc.acro)
tetrapodtraits.herp.known.karyo <- tetrapodtraits.herp.known.karyo %>% mutate(perc.acro.binned = cut(perc.acro, breaks = c(0, 0.01, 0.25, 0.4), include.lowest=TRUE))
tetrapodtraits.herp.known.karyo$perc.acro.binned[is.na(tetrapodtraits.herp.known.karyo$perc.acro.binned)] <- 0
tetrapodtraits.herp.known.karyo
```

```{r}
ggplot(tetrapodtraits.herp.known.karyo, aes(x=perc.acro.binned, color=sex)) +
  geom_bar(fill="white", position="dodge", alpha=0.5)
```

```{r}
tetrapodtraits.phyllostom <- add_column(tetrapodtraits.phyllostom, perc.acro = "")
tetrapodtraits.phyllostom$perc.acro <- karyotype.data[match(tetrapodtraits.phyllostom$Scientific.Name, karyotype.data$Species),5]
tetrapodtraits.phyllostom <- tetrapodtraits.phyllostom[!sapply(tetrapodtraits.phyllostom, function(x) all(x == ""))]
tetrapodtraits.phyllostom
```

```{r}
tetrapodtraits.phyllostom$perc.acro[which(tetrapodtraits.phyllostom$Scientific.Name %in% c("Carollia perspicillata", "Ariteus flavescens", "Artibeus fimbriatus", "Artibeus jamaicensis", "Artibeus lituratus", "Artibeus obscurus", "Artibeus planirostris", "Dermanura cinerea", "Dermanura phaeotis", "Dermanura tolteca", "Mesophylla macconnelli", "Phyllops falcatus", "Uroderma magnirostrum"))] <- c(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0.16666667)
tetrapodtraits.phyllostom$perc.acro
```

```{r}
tetrapodtraits.phyllo.known.karyo <- tetrapodtraits.phyllostom[!is.na(tetrapodtraits.phyllostom$perc.acro),]
tetrapodtraits.phyllo.known.karyo
```

```{r}
ggplot(tetrapodtraits.phyllo.known.karyo, aes(fill=sex, x=perc.acro)) +
  geom_bar(position = "dodge")
```

```{r}
tetrapodtraits.sorex <- add_column(tetrapodtraits.sorex, perc.acro = "")
tetrapodtraits.sorex$perc.acro <- karyotype.data[match(tetrapodtraits.sorex$Scientific.Name, karyotype.data$Species),5]
tetrapodtraits.sorex <- tetrapodtraits.sorex[!sapply(tetrapodtraits.sorex, function(x) all(x == ""))]
tetrapodtraits.sorex
```


```{r}
tetrapodtraits.sorex$perc.acro[which(tetrapodtraits.sorex$Scientific.Name %in% c("Sorex antinorii", "Sorex maritimensis", "Sorex satunini"))] <- c(0.33333333, 0.71428571, 0.33333333)
tetrapodtraits.sorex$perc.acro
```

```{r}
tetrapodtraits.sorex.known.karyo <- tetrapodtraits.sorex[!is.na(tetrapodtraits.sorex$perc.acro),]
tetrapodtraits.sorex.known.karyo
```

```{r}
ggplot(tetrapodtraits.sorex.known.karyo, aes(x=perc.acro, color=sex)) +
  geom_bar(fill="white", alpha=0.5)
```

```{r}
log_reg <- c()
y <- tetrapodtraits.herp.known.karyo$perc.acro
names(y) <- unlist(tetrapodtraits.herp.known.karyo$Scientific.Name)
group <- factor(tetrapodtraits.herp.known.karyo$XYorvariant)
names(group) <- unlist(tetrapodtraits.herp.known.karyo$Scientific.Name) 
test <- phyloglm(group ~ y, phy = herpestidae.tree, data = tetrapodtraits.herp.known.karyo, method = c("logistic_MPLE"), boot = 1000)
log_reg <- rbind(log_reg, summary(test))
print(test$alphaWarn)
print(log_reg[1,2])
```

```{r}
log_reg <- c()
y <- tetrapodtraits.phyllo.known.karyo$perc.acro
names(y) <- unlist(tetrapodtraits.phyllo.known.karyo$Scientific.Name)
group <- factor(tetrapodtraits.phyllo.known.karyo$XYorvariant)
names(group) <- unlist(tetrapodtraits.phyllo.known.karyo$Scientific.Name) 
test <- phyloglm(group ~ y, phy = phyllostom.tree, data = tetrapodtraits.phyllo.known.karyo, method = c("logistic_MPLE"), boot = 1000)
log_reg <- rbind(log_reg, summary(test))
print(test$alphaWarn)
print(log_reg[1,2])
```

```{r}
log_reg <- c()
y <- tetrapodtraits.sorex.known.karyo$perc.acro
names(y) <- unlist(tetrapodtraits.sorex.known.karyo$Scientific.Name)
group <- factor(tetrapodtraits.sorex.known.karyo$XYorvariant)
names(group) <- unlist(tetrapodtraits.sorex.known.karyo$Scientific.Name) 
test <- phyloglm(group ~ y, phy = sorex.tree, data = tetrapodtraits.sorex.known.karyo, method = c("logistic_MPLE"), boot = 1000)
log_reg <- rbind(log_reg, summary(test))
print(test$alphaWarn)
print(log_reg[1,2])
## plot trees w taxa only from the known.karyo datasets, add perc.acro for the tips?
```

```{r}
herp.sensiphy.percacro <- tree_phyglm(XYorvariant ~ perc.acro, data = tetrapodtraits.herp.known.karyo, phy = herp.multiphylo, n.tree = 1000)
summary(herp.sensiphy.percacro)
```

```{r}
phyllo.sensiphy.percacro <- tree_phyglm(XYorvariant ~ perc.acro, data = tetrapodtraits.phyllo.known.karyo, phy = phyllostom.multiphylo, n.tree = 1000)
summary(phyllo.sensiphy.percacro)
```

```{r}
sorex.sensiphy.acro <- tree_phyglm(XYorvariant ~ perc.acro, data = tetrapodtraits.sorex.known.karyo, phy = sorex.multiphylo, n.tree = 1000)
summary(sorex.sensiphy.acro)
```

```{r}
## png(filename = "herp_sensiphy_p.png", width = 4000, height = 2000, units="px", res = 300, bg = NA)
## stripchart(herp.sensiphy.p, main="sensiPhy p-values over 1000 trees, Herpestidae", xlab="Variables studied", ylab="P-values", method = "jitter", jitter = 0.1, vertical = TRUE, cex = 0.2)
## dev.off()
## cowplot on ggplot

## temp <- setNames(tetrapodtraits.herp.known.karyo$perc.acro, rownames(tetrapodtraits.herp.known.karyo))
## pdf("tempfile.pdf", width = 10, height = 10)
## plotTree.barplot(herp.tree.known.karyo, temp)
## dev.off()
## herp.tree.known.karyo <- keep.tip(herpestidae.tree, tetrapodtraits.herp.known.karyo$Scientific.Name)
## herp.matrix <- as.matrix(tetrapodtraits.herp.known.karyo, rownames = 4)[,88]
## plotTree.boxplot(herp.tree.known.karyo, herp.matrix)
## make these plots in color! play with the scaling of the actual bar plots (don't make them so big?)
## do the taxa in known.karyo include all variant species?
## dotTree() with no dots if no known karyo?
## plot all trees on same "canvas" (biggest tree taking up half of canvas, second tree taking up 1/3, etc)
```

```{r}
## herp.sensiphy.p <- list("RangeSize"=herp.sensiphy.rangesize$sensi.estimates$pval.estimate, 
                        ##"AbsoluteLat"=herp.sensiphy.abslat$sensi.estimates$pval.estimate, 
                        ##"AnnuMeanTemp"=herp.sensiphy.annumeantemp$sensi.estimates$pval.estimate,
                        ##"AnnuPrecip"=herp.sensiphy.annuprecip$sensi.estimates$pval.estimate,
                        ##"TempSeasonality"=herp.sensiphy.tempseason$sensi.estimates$pval.estimate,
                        ##"PrecipSeasonality"=herp.sensiphy.precipseason$sensi.estimates$pval.estimate,
                        ##"Elevation"=herp.sensiphy.elevation$sensi.estimates$pval.estimate,
                        ##"% Acrocentric"=herp.sensiphy.percacro$sensi.estimates$pval.estimate)
```

```{r}
stripchart(herp.sensiphy.p, main="sensiPhy p-values over 1000 trees, Herpestidae", xlab="Variables studied", ylab="P-values", method = "jitter", jitter = 0.1, vertical = TRUE)
```

```{r}
phyllo.sensiphy.p <- list("RangeSize"=phyllostom.sensiphy.rangesize$sensi.estimates$pval.estimate, 
                        "AbsoluteLat"=phyllostom.sensiphy.abslat$sensi.estimates$pval.estimate, 
                        "AnnuMeanTemp"=phyllostom.sensiphy.annumeantemp$sensi.estimates$pval.estimate,
                        "AnnuPrecip"=phyllostom.sensiphy.annuprecip$sensi.estimates$pval.estimate,
                        "TempSeasonality"=phyllostom.sensiphy.tempseason$sensi.estimates$pval.estimate,
                        "PrecipSeasonality"=phyllostom.sensiphy.precipseason$sensi.estimates$pval.estimate,
                        "Elevation"=phyllostom.sensiphy.elevation$sensi.estimates$pval.estimate,
                        "% Acrocentric"=phyllo.sensiphy.percacro$sensi.estimates$pval.estimate)
```

```{r}
phyllo.sensiphy.p.trunc <- list("RangeSize"=phyllostom.sensiphy.rangesize$sensi.estimates$pval.estimate, 
                        "AbsoluteLat"=phyllostom.sensiphy.abslat$sensi.estimates$pval.estimate, 
                        "AnnuMeanTemp"=phyllostom.sensiphy.annumeantemp$sensi.estimates$pval.estimate,
                        "PrecipSeasonality"=phyllostom.sensiphy.precipseason$sensi.estimates$pval.estimate)
```

```{r}
##shrew.sensiphy.p <- list("RangeSize"=sorex.sensiphy.rangesize$sensi.estimates$pval.estimate, 
                        ##"AbsoluteLat"=sorex.sensiphy.abslat$sensi.estimates$pval.estimate, 
                        ##"AnnuMeanTemp"=sorex.sensiphy.annumeantemp$sensi.estimates$pval.estimate,
                        ##"AnnuPrecip"=sorex.sensiphy.annuprecip$sensi.estimates$pval.estimate,
                        ##"TempSeasonality"=sorex.sensiphy.tempseason$sensi.estimates$pval.estimate,
                        ##"PrecipSeasonality"=sorex.sensiphy.precipseason$sensi.estimates$pval.estimate,
                        ##"Elevation"=sorex.sensiphy.elevation$sensi.estimates$pval.estimate,
                        ##"% Acrocentric"=sorex.sensiphy.acro$sensi.estimates$pval.estimate)
```

```{r}
tetrapodtraits.herp.known.karyo <- add_column(tetrapodtraits.herp.known.karyo, haploid = "")
tetrapodtraits.herp.known.karyo$haploid <- karyotype.data[match(tetrapodtraits.herp.known.karyo$Scientific.Name, karyotype.data$Species),2]
tetrapodtraits.herp.known.karyo <- tetrapodtraits.herp.known.karyo[!sapply(tetrapodtraits.herp.known.karyo, function(x) all(x == ""))]
tetrapodtraits.herp.known.karyo
```

```{r}
tetrapodtraits.herp.known.karyo$haploid[which(tetrapodtraits.herp.known.karyo$Scientific.Name %in% c("Bdeogale nigripes", "Ichneumia albicauda", "Suricata suricatta"))] <- c(18, 18, 18)
tetrapodtraits.herp.known.karyo$haploid
```

```{r}
tetrapodtraits.phyllo.known.karyo <- add_column(tetrapodtraits.phyllo.known.karyo, haploid = "")
tetrapodtraits.phyllo.known.karyo$haploid <- karyotype.data[match(tetrapodtraits.phyllo.known.karyo$Scientific.Name, karyotype.data$Species),2]
tetrapodtraits.phyllo.known.karyo <- tetrapodtraits.phyllo.known.karyo[!sapply(tetrapodtraits.phyllo.known.karyo, function(x) all(x == ""))]
tetrapodtraits.phyllo.known.karyo
```

```{r}
tetrapodtraits.phyllo.known.karyo$haploid[which(tetrapodtraits.phyllo.known.karyo$Scientific.Name %in% c("Ariteus flavescens", "Artibeus fimbriatus", "Artibeus jamaicensis", "Artibeus lituratus", "Artibeus obscurus", "Artibeus planirostris", "Carollia perspicillata", "Dermanura cinerea", "Dermanura phaeotis", "Dermanura tolteca", "Mesophylla macconnelli", "Phyllops falcatus", "Uroderma magnirostrum"))] <- c(15, 15, 15, 15, 15, 15, 10, 15, 15, 15, 11, 15, 18)
tetrapodtraits.phyllo.known.karyo$haploid
## subtract 1 from haploid count of all species
```

```{r}
tetrapodtraits.sorex.known.karyo <- add_column(tetrapodtraits.sorex.known.karyo, haploid = "")
tetrapodtraits.sorex.known.karyo$haploid <- karyotype.data[match(tetrapodtraits.sorex.known.karyo$Scientific.Name, karyotype.data$Species),2]
tetrapodtraits.sorex.known.karyo <- tetrapodtraits.sorex.known.karyo[!sapply(tetrapodtraits.sorex.known.karyo, function(x) all(x == ""))]
tetrapodtraits.sorex.known.karyo
## subtract 1 from haploid count of all species
```

```{r}
tetrapodtraits.sorex.known.karyo$haploid[which(tetrapodtraits.sorex.known.karyo$Scientific.Name %in% c("Sorex antinorii", "Sorex maritimensis", "Sorex satunini"))] <- c(12, 14, 12)
tetrapodtraits.sorex.known.karyo$haploid
## subtract 1 from haploid count of all species
## for parameter estimates, record estimates because these are slow (variable w on the example script), save final output
## new column with haploid count - 1 + if loop, where if column with specific name exists, delete it
## drop Sorex kozlovi
```

```{r}
herp.auto.hap <- tetrapodtraits.herp.known.karyo[,89] - 1
tetrapodtraits.herp.known.karyo <- add_column(tetrapodtraits.herp.known.karyo, auto.hap = "")
tetrapodtraits.herp.known.karyo$auto.hap <- herp.auto.hap
tetrapodtraits.herp.known.karyo <- tetrapodtraits.herp.known.karyo[!sapply(tetrapodtraits.herp.known.karyo, function(x) all(x == ""))]
tetrapodtraits.herp.known.karyo$auto.hap
```

```{r}
phyllo.auto.hap <- tetrapodtraits.phyllo.known.karyo[,87] - 1
tetrapodtraits.phyllo.known.karyo <- add_column(tetrapodtraits.phyllo.known.karyo, auto.hap = "")
tetrapodtraits.phyllo.known.karyo$auto.hap <- phyllo.auto.hap
tetrapodtraits.phyllo.known.karyo <- tetrapodtraits.phyllo.known.karyo[!sapply(tetrapodtraits.phyllo.known.karyo, function(x) all(x == ""))]
tetrapodtraits.phyllo.known.karyo$auto.hap
```

```{r}
sorex.auto.hap <- tetrapodtraits.sorex.known.karyo[,87] - 1
tetrapodtraits.sorex.known.karyo <- add_column(tetrapodtraits.sorex.known.karyo, auto.hap = "")
tetrapodtraits.sorex.known.karyo$auto.hap <- sorex.auto.hap
tetrapodtraits.sorex.known.karyo <- tetrapodtraits.sorex.known.karyo[!sapply(tetrapodtraits.sorex.known.karyo, function(x) all(x == ""))]
tetrapodtraits.sorex.known.karyo$auto.hap
```

```{r}
herp.tree.known.karyo <- keep.tip(herpestidae.tree, tetrapodtraits.herp.known.karyo$Scientific.Name)
herp.tree.known.karyo
```

```{r}
phyllo.tree.known.karyo <- keep.tip(phyllostom.tree, tetrapodtraits.phyllo.known.karyo$Scientific.Name)
phyllo.tree.known.karyo
```

```{r}
sorex.tree.known.karyo <- keep.tip(sorex.tree, tetrapodtraits.sorex.known.karyo$Scientific.Name)
sorex.tree.known.karyo
```
```{r}
chromeplus.herp.df <- tetrapodtraits.herp.known.karyo %>% select(Scientific.Name, auto.hap, XYorvariant)
chromeplus.herp.df
```

```{r}
chromeplus.herp.matrix <- datatoMatrix(chromeplus.herp.df, hyper = T)
chromeplus.herp.matrix
```

```{r}
iter <- 1000
herp.tree.known.karyo <- force.ultrametric(herp.tree.known.karyo)
herp.lk.mk <- make.mkn(tree = herp.tree.known.karyo, states = chromeplus.herp.matrix, k = ncol(chromeplus.herp.matrix), strict=F, control=list(method="ode"))
herp.con.lk.mk<-constrainMkn(data=chromeplus.herp.matrix, lik=herp.lk.mk, hyper=T, 
                        polyploidy=F, verbose=F, 
                        constrain=list(drop.demi=T, drop.poly=T))
temp <- mcmc(herp.con.lk.mk, x.init=runif(6,0,10), w=1, nsteps=iter/100)
w <- diff(sapply(temp[1:6], quantile, c(.05, .95)))
```

```{r}
consensus.results <- c(0.1183770, 0.3903215, 0.1183770, 0.1183770)
png(filename = "phyllo_sensiphy_p.png", width = 3000, height = 1500, units="px", res = 300, bg = NA)
stripchart(phyllo.sensiphy.p.trunc, main="Differences in range of statistical outcomes over 1000 trees compared to statistical outcome of single tree in Phyllostomidae", xlab="Eco-environmental variables studied", ylab="P-values", method = "jitter", jitter = 0.1, vertical = TRUE, cex = 0.2)
points(consensus.results, col = "red", pch = 7, cex = 1.5, bg = 2, lwd = 2)
dev.off()
```





